### ISSUE TITLE
Recieving TimeoutExceptions when calling URLs with DNS names

### ISSUE THREAD

**Issue by rGunti** [dotnet/core#8151]
### Description

I have resurrected development on a personal project after a short hiatus. After bringing it back online, I've been facing strange timeout issues. Upon closer investigation I was able to narrow it down to Timeouts when calling for a web resource. To make sure it is not a fluke, I created a test console project and was able to reproduce the issue (see below).

Basically, when calling any URL that contains a domain name, it will fail with a timeout. When an IP address is provided in the URL, it works without issues. This only occurs on my Pop!_OS (Ubuntu) machine, on Windows I don't face this problem. Additionally, if I run this project inside a Docker container, it also seems to work just fine.

I've tried reinstalling the dotnet SDK several times, rebooted the machine multiple times as well. No success.

This issue seems to only affect dotnet, as any other process on the system works just fine when connecting to the internet.

#### Source Code:

```cs
// DoesItWork.cs
using System.Net;

namespace Playground;

public static class DoYouWork
{
    public static void Test(string url)
    {
        Console.Write($"Testing {url,-60}: ");
        try
        {
            CheckForInternetConnection(url: url);
            Console.WriteLine("PASS");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"FAIL! {ex.GetType()}: {ex.Message}");
        }
    }

    public static bool CheckForInternetConnection(string url)
    {
        // Authors note: I've seen the same results with other dotnet methods of calling web resources (i.e. HttpClient, async methods, ...).
        var request = (HttpWebRequest)WebRequest.Create(url);
        request.KeepAlive = false;
        request.Timeout = 10000;
        using (var response = (HttpWebResponse)request.GetResponse())
            return true;
    }
}

// Program.cs
DoYouWork.Test(url: "http://1.1.1.1");
DoYouWork.Test(url: "https://1.1.1.1");
DoYouWork.Test(url: "http://google.com");
DoYouWork.Test(url: "https://google.com");
DoYouWork.Test(url: "http://www.gstatic.com/generate_204");
DoYouWork.Test(url: "https://www.gstatic.com/generate_204");
```

Dockerfile:
```
FROM mcr.microsoft.com/dotnet/sdk:7.0 AS build
COPY ./ /src/

RUN dotnet restore /src/Playground/Playground.csproj
RUN dotnet build /src/Playground/Playground.csproj
RUN dotnet publish -o /out /src/Playground/Playground.csproj

FROM mcr.microsoft.com/dotnet/runtime:7.0 AS runtime
COPY --from=build /out /app
WORKDIR /app
ENTRYPOINT [ "dotnet", "Playground.dll" ]
```

Output, when run locally:
```
$ dotnet run --project Playground/Playground.csproj 
Testing http://1.1.1.1                                              : PASS
Testing https://1.1.1.1                                             : PASS
Testing http://google.com                                           : FAIL! System.Net.WebException: The operation has timed out.
Testing https://google.com                                          : FAIL! System.Net.WebException: The operation has timed out.
Testing http://www.gstatic.com/generate_204                         : FAIL! System.Net.WebException: The operation has timed out.
Testing https://www.gstatic.com/generate_204                        : FAIL! System.Net.WebException: The operation has timed out.
```

### Configuration

```
$ dotnet --info
.NET SDK:
 Version:   7.0.102
 Commit:    4bbdd14480

Runtime Environment:
 OS Name:     pop
 OS Version:  22.04
 OS Platform: Linux
 RID:         linux-x64
 Base Path:   /usr/share/dotnet/sdk/7.0.102/

Host:
  Version:      7.0.2
  Architecture: x64
  Commit:       d037e070eb

.NET SDKs installed:
  6.0.405 [/usr/share/dotnet/sdk]
  7.0.102 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.13 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.13 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/usr/lib/dotnet]

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

### Regression?

don't know

### Other information

- Is dotnet resolving domain names differently than the rest of the OS? If so is there a way to influence name resolution from the outside?

**Comment by carlossanlop** [comment]
@dotnet/ncl any ideas why this could be happening?

**Comment by ManickaP** [comment]
Looks like name resolution issue in the docker image. Does `nslookup ...` work inside the image? Does the error reproduces on other machines (e.g. outside of docker)?
BTW, `WebRequest` is not recommended to be used in .NET Core and newer (it's kept only for backward compatibility). Always use `HttpClient` and always prefer `async` methods.

**Comment by rGunti** [comment]
Result of `nslookup`:

```
$ nslookup google.com
Server:		127.0.0.53
Address:	127.0.0.53#53

Non-authoritative answer:
Name:	google.com
Address: 142.250.184.206
Name:	google.com
Address: 2a00:1450:4001:830::200e

$ nslookup gstatic.com   
Server:		127.0.0.53
Address:	127.0.0.53#53

Non-authoritative answer:
Name:	gstatic.com
Address: 142.250.186.35
Name:	gstatic.com
Address: 2a00:1450:4001:827::2003
```

I am well aware that `WebRequest` is deprecated but I've seen this issue first when setting up JWT Bearer authentication using `.AddJwtBearer()`. When initialized, it tries to contact a `.well-known` URL of the JWT Authority, which ultimately fails. The code above is just a sample I [grabbed of StackOverflow](https://stackoverflow.com/questions/2031824/what-is-the-best-way-to-check-for-internet-connectivity-using-net) for testing purposes and modified it a bit later.

Regardless, I replaced the implementation with `HttpClient`, just to be sure. But no real change:
```cs
namespace Playground;

public static class DoYouWork
{
    private static HttpClient Client = new();

    static DoYouWork()
    {
        Client.Timeout = TimeSpan.FromSeconds(10);
    }
    
    public static async Task Test(string url)
    {
        Console.Write($"Testing {url,-60}: ");
        try
        {
            await CheckForInternetConnection(url: url);
            Console.WriteLine("PASS");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"FAIL! {ex.GetType()}: {ex.Message}");
        }
    }

    private static async Task CheckForInternetConnection(string url)
    {
        HttpResponseMessage response = await Client.GetAsync(url);
        response.EnsureSuccessStatusCode();
    }
}
```

Output:
```
Testing http://1.1.1.1                                              : PASS
Testing https://1.1.1.1                                             : PASS
Testing http://google.com                                           : FAIL! System.Threading.Tasks.TaskCanceledException: The request was canceled due to the configured HttpClient.Timeout of 10 seconds elapsing.
Testing https://google.com                                          : FAIL! System.Threading.Tasks.TaskCanceledException: The request was canceled due to the configured HttpClient.Timeout of 10 seconds elapsing.
Testing http://www.gstatic.com/generate_204                         : FAIL! System.Threading.Tasks.TaskCanceledException: The request was canceled due to the configured HttpClient.Timeout of 10 seconds elapsing.
Testing https://www.gstatic.com/generate_204                        : FAIL! System.Threading.Tasks.TaskCanceledException: The request was canceled due to the configured HttpClient.Timeout of 10 seconds elapsing.
```

I didn't mention it above, but it also seems to be SDK-agnostic. Doesn't mater if the project is `net6.0` or `net7.0`, the output is the same. And again: It seems to be just this one machine, which would hint at some misconfiguration in the OS, but it seems quite curious that only dotnet is affected and any other application on the system is perfectly fine.

**Comment by rGunti** [comment]
After a bit more experimentation, I have figured out the issue: **IPv6**

While going through other options, I remembered that I had some issues with IPv6 some time ago (thought it should be fixed now). I then followed it up by disabling IPv6 on my system by running this command:

```sh
sudo -- sh -c 'sysctl -w net.ipv6.conf.all.disable_ipv6=1 && sysctl -w net.ipv6.conf.default.disable_ipv6=1'
```

I then restarted my IDE, just to be sure and I got an instant `PASS` on all my test addresses. To confirm this, I looked for domains that only resolve in IPv4 and IPv6 respectively (`ipv4.google.com` and `ipv6.google.com` come to mind) and tested those with IPv6 enabled (using `sudo -- sh -c 'sysctl -w net.ipv6.conf.all.disable_ipv6=0 && sysctl -w net.ipv6.conf.default.disable_ipv6=0'`). Interestingly I don't get a timeout anymore after this exercise, I get a `HttpRequestException: No route to host (ipv6.google.com:443)` while all other addresses are just slow to resolve but `PASS` eventually.

I am not entierly sure why this issue doesn't pop up anymore, but it is IPv6 related. I was also able to confirm this by `ping`ing an IPv6 address.

```sh
$ ping 2606:4700:4700::1111  # <= this is 1.1.1.1 for IPv6
PING 2606:4700:4700::1111(2606:4700:4700::1111) 56 data bytes
^C
--- 2606:4700:4700::1111 ping statistics ---
42 packets transmitted, 0 received, 100% packet loss, time 41964ms
```

Anyway, sorry for taking up your time with my misconfigured OS. I hope this helps anyone else experiencing a similar problem in the future.
